File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    let min=prices[0];
    let maxprofit=0;
    for(let i=1;i<prices.length;i++){
        if(prices[i]<min){
            min=prices[i];
        }
        let profit=prices[i]-min;
        maxprofit=Math.max(maxprofit,profit);
    }
    return maxprofit;
    
};

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n=prices.size();
        vector<int>suffix(n);
        suffix[n-1]=prices[n-1];
        for(int i=n-2;i>=0;i--){
            suffix[i]=max(prices[i],suffix[i+1]);
        }
        int profit=0;
        for(int i=0;i<n-1;i++){
            profit=max(profit,suffix[i+1]-prices[i]);
        }
        return profit;
    }
};